components/eam/src/physics/cam/modal_aer_opt.F90

```diff
@@ -1302,81 +1297,54 @@ subroutine modal_aero_lw(list_idx, dt, state, pbuf, tauxar)
 
-      dgnumwet => dgnumwet_m(:,:,m)
-      qaerwat  => qaerwat_m(:,:,m)
+   ! loop over all aerosol modes
+   do mm = 1, nmodes
+
+      dgnumwet => dgnumwet_m(:,:,mm)
+      qaerwat  => qaerwat_m(:,:,mm)
 
       ! get mode properties
-      call rad_cnst_get_mode_props(list_idx, m, sigmag=sigma_logr_aer, refrtablw=refrtablw , &
+      call rad_cnst_get_mode_props(list_idx, mm, sigmag=sigma_logr_aer, refrtablw=refrtablw , &
          refitablw=refitablw, absplw=absplw)
 
       ! get mode info
-      call rad_cnst_get_info(list_idx, m, nspec=nspec)
+      call rad_cnst_get_info(list_idx, mm, nspec=nspec)
 
       ! calc size parameter for all columns
-      ! this is the same calculation that's done in modal_size_parameters, but there
-      ! some intermediate results are saved and the chebyshev polynomials are stored
-      ! in a array with different index order.  Could be unified.
-      do k = top_lev, pver
-         do i = 1, ncol
-            alnsg_amode = log( sigma_logr_aer )
-            ! convert from number diameter to surface area
-            xrad(i) = log(0.5_r8*dgnumwet(i,k)) + 2.0_r8*alnsg_amode*alnsg_amode
-            ! normalize size parameter
-            xrad(i) = max(xrad(i), xrmin)
-            xrad(i) = min(xrad(i), xrmax)
-            xrad(i) = (2*xrad(i)-xrmax-xrmin)/(xrmax-xrmin)
-            ! chebyshev polynomials
-            cheby(1,i,k) = 1.0_r8
-            cheby(2,i,k) = xrad(i)
-            do nc = 3, ncoef
-               cheby(nc,i,k) = 2.0_r8*xrad(i)*cheby(nc-1,i,k)-cheby(nc-2,i,k)
-            end do
-         end do
-      end do
+      ! FORTRAN refactoring: ismethod is tempararily used to ensure BFB test. 
+      ! can be removed when porting to C++
+      call modal_size_parameters(ncol, sigma_logr_aer, dgnumwet, & ! in
+                                 radsurf, logradsurf, cheby, ismethod2=.true.) 
+
+      allocate(volf_l(ncol,nspec),stat=istat)
+      if (istat /= 0) call endrun("Unable to allocate volf_l: "//errmsg(__FILE__,__LINE__) )
+      allocate(specdens_l(nspec),stat=istat)
+      if (istat /= 0) call endrun("Unable to allocate specdens_l: "//errmsg(__FILE__,__LINE__) )
+      allocate(specrefindex_l(nspec,nlwbands),stat=istat)
+      if (istat /= 0) call endrun("Unable to allocate specrefindex_l: "//errmsg(__FILE__,__LINE__) )
 
       do ilw = 1, nlwbands
 
-         do k = top_lev, pver
-
-            ! form bulk refractive index. Use volume mixing for infrared
-            crefin(:ncol) = (0._r8, 0._r8)
-            dryvol(:ncol) = 0._r8
-
-            ! aerosol species loop
-            do l = 1, nspec
-               call rad_cnst_get_aer_mmr(list_idx, m, l, 'a', state, pbuf, specmmr)
-               call rad_cnst_get_aer_props(list_idx, m, l, density_aer=specdens, &
+         do kk = top_lev, pver
+            ! get aerosol properties and save for each species
+            do ll = 1, nspec
+               call rad_cnst_get_aer_mmr(list_idx, mm, ll, 'a', state, pbuf, specmmr)
```

Comment:
The `mmr` can be obtained from `state%q (index)`.`index` can be obtained using `lmassptr_amode` array from modal_aero_data.